src/agent/onefuzz/src/monitor.rs

```diff
@@ -2,36 +2,33 @@
 // Licensed under the MIT License.
 
 use std::path::PathBuf;
-use std::sync::{self, mpsc::Receiver as SyncReceiver};
-use std::time::Duration;
 
-use anyhow::Result;
-use notify::{DebouncedEvent, Watcher};
+use anyhow::{format_err, Result};
+use notify::{Event, EventKind, Watcher};
 use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver};
-use tokio::task::{self, JoinHandle};
 
 pub struct DirectoryMonitor {
     dir: PathBuf,
-    notify_events: UnboundedReceiver<DebouncedEvent>,
+    notify_events: UnboundedReceiver<notify::Result<Event>>,
     watcher: notify::RecommendedWatcher,
 }
 
 impl DirectoryMonitor {
-    pub fn new(dir: impl Into<PathBuf>) -> Self {
+    pub fn new(dir: impl Into<PathBuf>) -> Result<Self> {
         let dir = dir.into();
-        let (notify_sender, notify_receiver) = sync::mpsc::channel();
-        let delay = Duration::from_millis(100);
-        let watcher = notify::watcher(notify_sender, delay).unwrap();
-
-        // We can drop the thread handle, and it will continue to run until it
-        // errors or we drop the async receiver.
-        let (notify_events, _handle) = into_async(notify_receiver);
-
-        Self {
+        let (sender, notify_events) = unbounded_channel();
+        let event_handler = move |event_or_err| {
+            // A send error only occurs when the channel is closed. No remedial
+            // action is needed (or possible), so ignore it.
+            let _ = sender.send(event_or_err);
```

Comment:
We ignore `Sender` errors on this line, but are still sending `Watcher` errors, so the monitor can report them and shutdown as nicely as possible.